api/listener.go

```diff
@@ -65,31 +79,69 @@ func (cl *chainListener) Stop() error {
 // ReceiveBlock handles the block
 func (cl *chainListener) ReceiveBlock(blk *block.Block) error {
 	// pass the block to every responder
-	cl.streamMap.Range(func(key, _ interface{}) error {
-		r, ok := key.(Responder)
+	cl.streamMap.Range(func(_, value interface{}) error {
+		r, ok := value.(apitypes.Responder)
 		if !ok {
-			log.L().Error("streamMap stores a key which is not a Responder")
-			return errorKeyIsNotResponder
+			log.L().Error("streamMap stores a value which is not a Responder")
+			return errorValueIsNotResponder
 		}
 		err := r.Respond(blk)
 		if err != nil {
 			log.L().Error("responder failed to process block", zap.Error(err))
-
 		}
 		return err
 	})
 	return nil
 }
 
 // AddResponder adds a new responder
-func (cl *chainListener) AddResponder(r Responder) error {
-	_, loaded := cl.streamMap.Get(r)
-	if loaded {
-		return errorResponderAdded
-	}
+func (cl *chainListener) AddResponder(responder apitypes.Responder) (listenerID string, err error) {
 	if cl.streamMap.Count() >= cl.capacity {
-		return errorCapacityReached
+		err = errorCapacityReached
+		return
 	}
-	cl.streamMap.Set(r, struct{}{})
-	return nil
+	i := 0
+	for ; i < idRetry; i++ {
+		listenerID = cl.idGenerator.newID()
+		if _, exist := cl.streamMap.Get(listenerID); !exist {
+			break
+		}
+	}
```

Comment:
such that we can prevent racing issue here

Reply:
`streamMap` is thread safe. What is the racing issue here?